### PR TITLE
Add GetReference and TryGetArray to MemoryMarshal

### PR DESCRIPTION
**Pull request by ahsonkhan**
Sets up the ground work for:
https://github.com/dotnet/corefx/issues/25412
https://github.com/dotnet/corefx/issues/25615

Test updates and changes in corefx ~to follow after.~ https://github.com/dotnet/corefx/pull/25789

Following the staging plan from here: https://github.com/dotnet/corefx/issues/23881#issuecomment-343767740

1. Add MemoryExtensions.GetReference/TryGetArray
2. Convert all uses of DangerousGetPinnableReference/DangerousTryGetArray in coreclr, corefx, corert, corefxlab, aspnet, ... to MemoryExtensions.GetReference
3. Change DangerousGetPinnableReference to whatever we like to make it fit the pinning pattern and remove DangerousTryGetArray.

Doing it this way will avoid the need for complex staging or things being on the floor for extensive periods of time.

cc @jkotas, @stephentoub, @KrzysztofCwalina

### CODE REVIEW

File: src/mscorlib/shared/System/Span.cs [hunk 1]
```diff
@@ -409,6 +409,9 @@ public T[] ToArray()
         /// </summary>
         public static Span<T> Empty => default(Span<T>);
 
+        // This exposes the internal representation for Span-related apis use only.
+        internal ByReference<T> Pointer => _pointer;
```
Comment (jkotas) [inline src/mscorlib/shared/System/Span.cs]: This should return `ref T`. `ByReference<T>` should be used only for the fields.

Comment (ahsonkhan) [reply src/mscorlib/shared/System/Span.cs]: Why is that?

Wouldn't this make the implementation, essentially a duplicate, but internal, version of DangerousGetPinnableReference?

i.e.
```C#
internal ref T Reference => ref _pointer.Value;
```

Do we then just return `Reference` from MemoryMarshal.GetReference()?

File: src/mscorlib/shared/System/Runtime/InteropServices/MemoryMarshal.cs [hunk 2]
```diff
@@ -23,5 +24,32 @@ public static class MemoryMarshal
         /// </remarks>
         public static Memory<T> AsMemory<T>(ReadOnlyMemory<T> readOnlyMemory) =>
             Unsafe.As<ReadOnlyMemory<T>, Memory<T>>(ref readOnlyMemory);
+        
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (jkotas) [inline src/mscorlib/shared/System/Runtime/InteropServices/MemoryMarshal.cs]: Why is the AggresiveInlining necessary?

I know that you got into habit of marking everything with AggresiveInlining - I do not think it is a good habit to follow.

File: src/mscorlib/shared/System/Runtime/InteropServices/MemoryMarshal.cs [hunk 3]
```diff
@@ -23,5 +24,32 @@ public static class MemoryMarshal
...
+        public static ref T GetReference<T>(in Span<T> span) => ref span.Reference;
```
Comment (jkotas) [inline src/mscorlib/shared/System/Runtime/InteropServices/MemoryMarshal.cs]: Have you tried whether it is worth it to have `in` here? (It was discussed in the issue.)

File: src/mscorlib/shared/System/Runtime/InteropServices/MemoryMarshal.cs [hunk 4]
```diff
@@ -23,5 +24,32 @@ public static class MemoryMarshal
...
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ref readonly T GetReference<T>(in ReadOnlySpan<T> span) => ref span.Reference;
```
Comment (jkotas) [inline src/mscorlib/shared/System/Runtime/InteropServices/MemoryMarshal.cs]: I do not think the Reference property will get inlined here for the class case. We may want to make the field internal instead.

File: src/mscorlib/shared/System/Runtime/InteropServices/MemoryMarshal.cs [hunk 5]
```diff
@@ -23,5 +24,32 @@ public static class MemoryMarshal
...
+
+        public static bool TryGetArray<T>(ReadOnlyMemory<T> readOnlyMemory, out ArraySegment<T> arraySegment)
+        {
+            object obj = readOnlyMemory.GetObjectStartLength(out int index, out int length);
+            if (index < 0)
+            {
+                if (((OwnedMemory<T>)obj).TryGetArray(out var segment))
```
Comment (jkotas) [inline src/mscorlib/shared/System/Runtime/InteropServices/MemoryMarshal.cs]: Can this be just `return (((OwnedMemory<T>)obj).TryGetArray(out var arraySegment)` ?

File: src/mscorlib/shared/System/Runtime/InteropServices/MemoryMarshal.cs (hunk 2 above)
Comment (ahsonkhan) [reply src/mscorlib/shared/System/Runtime/InteropServices/MemoryMarshal.cs]: For this instance, I wanted to remain consistent with DangerousGetPinnableReference which already had it, presumable for good reason. I will try to run some experiments.

File: src/mscorlib/shared/System/Runtime/InteropServices/MemoryMarshal.cs (hunk 4 above)
Comment (ahsonkhan) [reply src/mscorlib/shared/System/Runtime/InteropServices/MemoryMarshal.cs]: The `ByReference<T> _pointer` field?

File: src/mscorlib/shared/System/Runtime/InteropServices/MemoryMarshal.cs (hunk 2 above)
Comment (jkotas) [reply src/mscorlib/shared/System/Runtime/InteropServices/MemoryMarshal.cs]: ok

File: src/mscorlib/shared/System/Runtime/InteropServices/MemoryMarshal.cs (hunk 5 above)
Comment (jkotas) [reply src/mscorlib/shared/System/Runtime/InteropServices/MemoryMarshal.cs]: I see - you need to add the local index.

File: src/mscorlib/shared/System/Runtime/InteropServices/MemoryMarshal.cs (hunk 4 above)
Comment (jkotas) [reply src/mscorlib/shared/System/Runtime/InteropServices/MemoryMarshal.cs]: Yes.

File: src/mscorlib/shared/System/Runtime/InteropServices/MemoryMarshal.cs (hunk 2 above)
Comment (ahsonkhan) [reply src/mscorlib/shared/System/Runtime/InteropServices/MemoryMarshal.cs]: Inlining in almost all cases made no difference.

The only case inlining helps is if the implementation is using `in` and the `Reference` property.

File: src/mscorlib/shared/System/Runtime/InteropServices/MemoryMarshal.cs (hunk 3 above)
Comment (ahsonkhan) [reply src/mscorlib/shared/System/Runtime/InteropServices/MemoryMarshal.cs]: Using `in` doesn't really help (at least for fast span). And if we don't inline the method, it actually hurts performance.

If I pass a Span using `in`, it seems to hurt performance in some cases:

```C#
[MethodImpl(MethodImplOptions.NoInlining)]
public static void CallTestMethodWithIn(Span<int> span)
{
    int temp = TestMethodWithIn(span);
}

[MethodImpl(MethodImplOptions.NoInlining)]
public static void CallTestMethodWithoutIn(Span<int> span)
{
    int temp = TestMethodWithoutIn(span);
}

[MethodImpl(MethodImplOptions.NoInlining)]
public static int TestMethodWithIn(in ReadOnlySpan<int> span)
{
    int val = 0;
    for(int i = 0; i < span.Length; i++)
        val ^= span[i];
    return val;
}

[MethodImpl(MethodImplOptions.NoInlining)]
public static int TestMethodWithoutIn(ReadOnlySpan<int> span)
{
    int val = 0;
    for (int i = 0; i < span.Length; i++)
        val ^= span[i];
    return val;
}
```
**Runtime (s):**
CallTestMethodWithoutIn: 6.3821768
CallTestMethodWithIn: 6.8055258

There is an extra compare and jump in the loop body if I pass the Span with `in`.
image

File: src/mscorlib/shared/System/Runtime/InteropServices/MemoryMarshal.cs (hunk 4 above)
Comment (ahsonkhan) [reply src/mscorlib/shared/System/Runtime/InteropServices/MemoryMarshal.cs]: Yes, it looks like without `AggressiveInlining` it doesn't get inlined.

**Comment by ahsonkhan** [comment]
I tried to toggle the implementation based on these 3 factors:
- AggressiveInlining
- Use of `in`
- Use `Reference` property or `_pointer` field

Most of these had negligible impact. The only one that made a difference was AggressiveInlining only if I passed the Span using `in`.
image

Given that these factors don't really affect the results (and using `in` requires `AggressiveInlining`), I will use this implementation:
```C#
GetReference(Span<T> span) => ref span._pointer.Value;
```

**Pull request merged**